oauthenticator/gitlab.py

```diff
@@ -144,12 +144,12 @@ class GitLabOAuthenticator(OAuthenticator):
 
         if self.gitlab_group_whitelist:
             is_group_specified = True
-            user_in_group = await self._check_group_whitelist(user_id, access_token)
+            user_in_group = await self._check_group_whitelist(user_id, username, access_token)
 
         # We skip project_id check if user is in whitelisted group.
         if self.gitlab_project_id_whitelist and not user_in_group:
             is_project_id_specified = True
-            user_in_project = await self._check_project_id_whitelist(user_id, access_token)
+            user_in_project = await self._check_project_id_whitelist(user_id, username, access_token)
 
```

Comment:
Are these two conditionals for group/project whitelists ordered deliberately already? If not, consider what is most likely to reduce the amount of requests. I think it is good as it is though as I imagine groups have a bigger user base within them and the potential API rate limiting is probably relating to the amount of requests rather than how much data passed back.

Reply:
Good question!

It's hard to say about ideal ordering, because it largely depends on which list is longer, which seems to be entirely up to the user. And then there's considerations like that larger groups first to increase chances of an early return

For JupyterHub installed as a GitLab-managed app on a Kubernetes cluster, one of these lists will generally be empty, and the other will include exactly the project or group associated with the cluster.

In any case, I think it's safe to say neither before nor after this MR would it be recommended to have more than a handful of whitelisted groups and projects, especially given the [10 requests per second per IP address](https://docs.gitlab.com/ee/user/gitlab_com/index.html#haproxy-api-throttle) limitation on GitLab.com

Reply:
Thinking more about deliberate ordering, I'm not sure if we should bother until we get actual user complaints. And the same for the rate limits.

Reply:
Actually, one idea that is simple enough not to require additional testing: We could loop over the whitelists in a random order (different order each time). That way, at least the problem has a chance of getting better on a refresh for legitimate users.

Reply:
I think the note in the readme should be sufficient for now